File: TrumpMobile/TrumpMobile/Utils/FirebaseOrderManager.swift
import FirebaseAuth
import FirebaseFirestore
import Foundation

class FirebaseOrderManager {
  static let shared = FirebaseOrderManager()
  private let db = Firestore.firestore()

  private init() {}

  // MARK: - Step Progress Helpers

  /// Save the user's progress for an order, updating currentStep and status.
  func saveStepProgress(
    userId: String,
    orderId: String,
    step: Int,
    data: [String: Any] = [:],
    completion: ((Result<Void, Error>) -> Void)? = nil
  ) {
    var payload: [String: Any] = data
    payload["userId"] = userId
    payload["currentStep"] = step
    // keep status "pending" to match existing model unless it's completed
    payload["status"] = payload["status"] ?? "pending"
    payload["updatedAt"] = FieldValue.serverTimestamp()

    db.collection("users").document(userId)
      .collection("orders").document(orderId)
      .setData(payload, merge: true) { error in
        if let error = error {
          completion?(.failure(error))
        } else {
          completion?(.success(()))
        }
      }
  }

  /// Mark an order as completed
  func markOrderCompleted(
    userId: String, orderId: String, completion: ((Result<Void, Error>) -> Void)? = nil
  ) {
    let data: [String: Any] = [
      "status": "completed",
      "updatedAt": FieldValue.serverTimestamp(),
    ]

    db.collection("users").document(userId)
      .collection("orders").document(orderId)
      .setData(data, merge: true) { error in
        if let error = error {
          completion?(.failure(error))
        } else {
          completion?(.success(()))
        }
      }
  }

  /// Fetch the latest in-progress order (status pending) for a user to resume
  func fetchLatestIncompleteOrder(
    for userId: String,
    completion: @escaping (Result<(orderId: String, currentStep: Int), Error>) -> Void
  ) {
    db.collection("users").document(userId)
      .collection("orders")
      .whereField("status", isEqualTo: "pending")
      .order(by: "updatedAt", descending: true)
      .limit(to: 1)
      .getDocuments { snapshot, error in
        if let error = error {
          completion(.failure(error))
          return
        }
        guard let doc = snapshot?.documents.first else {
          completion(
            .failure(
              NSError(
                domain: "FirebaseOrderManager", code: 404,
                userInfo: [NSLocalizedDescriptionKey: "No in-progress order found"])))
          return
        }
        let data = doc.data()
        let step = data["currentStep"] as? Int ?? 1
        completion(.success((orderId: doc.documentID, currentStep: step)))
      }
  }

  func fetchUserOrders(completion: @escaping ([TrumpOrder]) -> Void) {
    guard let userId = Auth.auth().currentUser?.uid else {
      print("DEBUG: Error in FirebaseOrderManager - No authenticated user")
      completion([])
      return
    }

    print("DEBUG: Fetching orders for userId: \(userId)")

    // Fix: Query the subcollection instead of root collection
    db.collection("users").document(userId).collection("orders")
      .order(by: "updatedAt", descending: true)
      .getDocuments { snapshot, error in
        if let error = error {
          print("DEBUG: Error fetching orders: \(error.localizedDescription)")
          completion([])
          return
        }

        let orders =
          snapshot?.documents.compactMap { doc -> TrumpOrder? in
            do {
              // Create TrumpOrder from document data
              let data = doc.data()

              // Map the document data to TrumpOrder properties
              guard let userId = data["userId"] as? String else { return nil }

              let order = TrumpOrder(
                id: doc.documentID,
                userId: userId,
                planName: data["planName"] as? String ?? "Trump Mobile Plan",
                amount: data["amount"] as? Double ?? 47.45,
                orderDate: (data["createdAt"] as? Timestamp)?.dateValue() ?? Date(),
                status: TrumpOrder.OrderStatus(rawValue: data["status"] as? String ?? "pending")
                  ?? .pending,
                billingCompleted: data["billingCompleted"] as? Bool ?? false,
                phoneNumber: data["selectedPhoneNumber"] as? String,
                simType: data["simType"] as? String ?? "Physical SIM",
                currentStep: data["currentStep"] as? Int
              )

              return order
            } catch {
              print("DEBUG: Error creating order from document: \(error)")
              return nil
            }
          } ?? []

        print("DEBUG: Successfully fetched \(orders.count) orders")
        completion(orders)
      }
  }

  func fetchCompletedOrders(completion: @escaping ([TrumpOrder]) -> Void) {
    guard let userId = Auth.auth().currentUser?.uid else {
      print("DEBUG: Error in FirebaseOrderManager - No authenticated user")
      completion([])
      return
    }

    print("DEBUG: Fetching completed orders for userId: \(userId)")

    // Fix: Query the subcollection with status filter
    db.collection("users").document(userId).collection("orders")
      .whereField("status", isEqualTo: "completed")
      .order(by: "updatedAt", descending: true)
      .getDocuments { snapshot, error in
        if let error = error {
          print("DEBUG: Error fetching completed orders: \(error.localizedDescription)")
          completion([])
          return
        }

        let orders =
          snapshot?.documents.compactMap { doc -> TrumpOrder? in
            do {
              let data = doc.data()

              guard let userId = data["userId"] as? String else { return nil }

              let order = TrumpOrder(
                id: doc.documentID,
                userId: userId,
                planName: data["planName"] as? String ?? "Trump Mobile Plan",
                amount: data["amount"] as? Double ?? 47.45,
                orderDate: (data["createdAt"] as? Timestamp)?.dateValue() ?? Date(),
                status: TrumpOrder.OrderStatus(rawValue: data["status"] as? String ?? "pending")
                  ?? .pending,
                billingCompleted: data["billingCompleted"] as? Bool ?? false,
                phoneNumber: data["selectedPhoneNumber"] as? String,
                simType: data["simType"] as? String ?? "Physical SIM",
                currentStep: data["currentStep"] as? Int
              )

              return order
            } catch {
              print("DEBUG: Error creating completed order from document: \(error)")
              return nil
            }
          } ?? []

        print("DEBUG: Successfully fetched \(orders.count) completed orders")
        completion(orders)
      }
  }

  // Alternative method to fetch orders by multiple statuses
  func fetchOrdersByStatus(
    _ statuses: [TrumpOrder.OrderStatus], completion: @escaping ([TrumpOrder]) -> Void
  ) {
    guard let userId = Auth.auth().currentUser?.uid else {
      print("DEBUG: Error in FirebaseOrderManager - No authenticated user")
      completion([])
      return
    }

    let statusStrings = statuses.map { $0.rawValue }

    // Fix: Query the subcollection with status filter
    db.collection("users").document(userId).collection("orders")
      .whereField("status", in: statusStrings)
      .order(by: "updatedAt", descending: true)
      .getDocuments { snapshot, error in
        if let error = error {
          print("DEBUG: Error fetching orders by status: \(error.localizedDescription)")
          completion([])
          return
        }

        let orders =
          snapshot?.documents.compactMap { doc -> TrumpOrder? in
            do {
              let data = doc.data()

              guard let userId = data["userId"] as? String else { return nil }

              let order = TrumpOrder(
                id: doc.documentID,
                userId: userId,
                planName: data["planName"] as? String ?? "Trump Mobile Plan",
                amount: data["amount"] as? Double ?? 47.45,
                orderDate: (data["createdAt"] as? Timestamp)?.dateValue() ?? Date(),
                status: TrumpOrder.OrderStatus(rawValue: data["status"] as? String ?? "pending")
                  ?? .pending,
                billingCompleted: data["billingCompleted"] as? Bool ?? false,
                phoneNumber: data["selectedPhoneNumber"] as? String,
                simType: data["simType"] as? String ?? "Physical SIM",
                currentStep: data["currentStep"] as? Int
              )

              return order
            } catch {
              print("DEBUG: Error creating order from document: \(error)")
              return nil
            }
          } ?? []

        print("DEBUG: Successfully fetched \(orders.count) orders with specified statuses")
        completion(orders)
      }
  }

  func deleteOrder(orderId: String, completion: @escaping (Bool) -> Void) {
    print("DEBUG: FirebaseOrderManager.deleteOrder called with orderId: \(orderId)")

    guard let userId = Auth.auth().currentUser?.uid else {
      print("DEBUG: Error in FirebaseOrderManager - No authenticated user")
      completion(false)
      return
    }

    print("DEBUG: Found userId: \(userId), proceeding with order deletion")

    // Use the existing FirebaseManager to delete the order
    FirebaseManager.shared.deleteOrder(userId: userId, orderId: orderId) { success, error in
      if success {
        print("DEBUG: FirebaseOrderManager - Successfully deleted order \(orderId)")
      } else {
        print(
          "DEBUG: FirebaseOrderManager - Failed to delete order: \(error?.localizedDescription ?? "unknown error")"
        )
      }
      completion(success)
    }
  }

  // MARK: - Fetch a single order document

  /// Fetch a single order document for the current authenticated user.
  /// - Parameters:
  ///   - orderId: The order document ID under users/{uid}/orders/{orderId}
  ///   - completion: Returns the raw document data on success
  func fetchOrderDocument(
    orderId: String,
    completion: @escaping (Result<[String: Any], Error>) -> Void
  ) {
    guard let userId = Auth.auth().currentUser?.uid else {
      completion(
        .failure(
          NSError(
            domain: "FirebaseOrderManager",
            code: 401,
            userInfo: [NSLocalizedDescriptionKey: "User not authenticated"]))
      )
      return
    }

    db.collection("users").document(userId)
      .collection("orders").document(orderId)
      .getDocument { snapshot, error in
        if let error = error {
          completion(.failure(error))
          return
        }
        guard let data = snapshot?.data() else {
          completion(
            .failure(
              NSError(
                domain: "FirebaseOrderManager",
                code: 404,
                userInfo: [NSLocalizedDescriptionKey: "Order not found"]))
          )
          return
        }
        completion(.success(data))
      }
  }
}

File: linkmobile/lib/services/firebase_order_manager.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import '../models/order_models.dart' as models;
import 'firebase_manager.dart';

class FirebaseOrderManager {
  static final FirebaseOrderManager _instance = FirebaseOrderManager._internal();
  factory FirebaseOrderManager() => _instance;
  FirebaseOrderManager._internal();

  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final FirebaseAuth _auth = FirebaseAuth.instance;
  final FirebaseManager _firebaseManager = FirebaseManager();

  Future<void> saveStepProgress({
    required String userId,
    required String orderId,
    required int step,
    Map<String, dynamic> data = const {},
  }) async {
    await _firebaseManager.updateOrder(userId, orderId, {
      'currentStep': step,
      'status': 'pending',
      ...data,
    });
  }

  Future<void> markOrderCompleted(String userId, String orderId) async {
    await _firebaseManager.updateOrder(userId, orderId, {
      'status': 'completed',
      'orderCompleted': true,
      'currentStep': 6,
    });
  }

  Future<void> markOrderPendingPortIn(String userId, String orderId) async {
    await _firebaseManager.updateOrder(userId, orderId, {
      'status': 'pending_port_in',
      'currentStep': 5,
    });
  }

  Future<Map<String, String>?> fetchLatestIncompleteOrder(String userId) async {
    try {
      // Fetch all pending and pending_port_in orders and sort in memory to avoid requiring a composite index
      final querySnapshot = await _firestore
          .collection('users')
          .doc(userId)
          .collection('orders')
          .where('status', whereIn: ['pending', 'pending_port_in'])
          .get();

      if (querySnapshot.docs.isEmpty) {
        return null;
      }

      // Sort by updatedAt in memory (descending) and get the latest one
      final sortedDocs = querySnapshot.docs.toList()
        ..sort((a, b) {
          final aUpdated = a.data()['updatedAt'] as Timestamp?;
          final bUpdated = b.data()['updatedAt'] as Timestamp?;
          if (aUpdated == null && bUpdated == null) return 0;
          if (aUpdated == null) return 1;
          if (bUpdated == null) return -1;
          return bUpdated.compareTo(aUpdated);
        });

      final doc = sortedDocs.first;
      final data = doc.data();
      return {
        'orderId': doc.id,
        'currentStep': (data['currentStep'] ?? 1).toString(),
      };
    } catch (e) {
      return null;
    }
  }

  Future<List<models.Order>> fetchUserOrders(String userId) async {
    try {
      final querySnapshot = await _firestore
          .collection('users')
          .doc(userId)
          .collection('orders')
          .orderBy('updatedAt', descending: true)
          .get();

      return querySnapshot.docs.map((doc) {
        final data = doc.data();
        return models.Order(
          id: doc.id,
          userId: userId,
          planName: data['planName'] ?? '',
          amount: (data['amount'] ?? 0).toDouble(),
          orderDate: (data['orderDate'] as Timestamp?)?.toDate() ?? DateTime.now(),
          status: models.OrderStatus.values.firstWhere(
            (e) => e.name == (data['status'] ?? 'pending'),
            orElse: () => models.OrderStatus.pending,
          ),
          billingCompleted: data['billingCompleted'] ?? false,
          phoneNumber: data['phoneNumber'],
          simType: data['simType'] ?? '',
          currentStep: data['currentStep'],
        );
      }).toList();
    } catch (e) {
      return [];
    }
  }

  Future<Map<String, dynamic>?> fetchOrderDocument(String userId, String orderId) async {
    return await _firebaseManager.getOrder(userId, orderId);
  }

  Future<List<models.Order>> fetchCompletedOrders(String userId) async {
    try {
      final querySnapshot = await _firestore
          .collection('users')
          .doc(userId)
          .collection('orders')
          .where('status', isEqualTo: 'completed')
          .orderBy('updatedAt', descending: true)
          .get();

      return querySnapshot.docs.map((doc) {
        final data = doc.data();

        // Get plan name and amount from order data
        final planName = data['planName'] ?? '';
        double amount;
        if (data['planPrice'] != null) {
          if (data['planPrice'] is int) {
            amount = (data['planPrice'] as int).toDouble();
          } else if (data['planPrice'] is double) {
            amount = data['planPrice'] as double;
          } else {
            amount = (data['amount'] ?? 0).toDouble();
          }
        } else {
          amount = (data['amount'] ?? 0).toDouble();
        }

        return models.Order(
          id: doc.id,
          userId: userId,
          planName: planName,
          amount: amount,
          orderDate: (data['orderDate'] as Timestamp?)?.toDate() ?? DateTime.now(),
          status: models.OrderStatus.values.firstWhere(
            (e) => e.name == (data['status'] ?? 'pending'),
            orElse: () => models.OrderStatus.pending,
          ),
          billingCompleted: data['billingCompleted'] ?? false,
          phoneNumber: data['phoneNumber']?.toString(),
          simType: data['simType']?.toString() ?? '',
          currentStep: data['currentStep'] as int?,
        );
      }).toList();
    } catch (e) {
      print('DEBUG: Error fetching completed orders: $e');
      return [];
    }
  }

  // Alternative method to fetch orders by multiple statuses
  Future<List<models.Order>> fetchOrdersByStatus(
    String userId,
    List<models.OrderStatus> statuses,
  ) async {
    try {
      final statusStrings = statuses.map((e) => e.name).toList();

      final querySnapshot = await _firestore
          .collection('users')
          .doc(userId)
          .collection('orders')
          .where('status', whereIn: statusStrings)
          .orderBy('updatedAt', descending: true)
          .get();

      return querySnapshot.docs.map((doc) {
        final data = doc.data();

        // Get plan name and amount from order data
        final planName = data['planName'] ?? '';
        double amount;
        if (data['planPrice'] != null) {
          if (data['planPrice'] is int) {
            amount = (data['planPrice'] as int).toDouble();
          } else if (data['planPrice'] is double) {
            amount = data['planPrice'] as double;
          } else {
            amount = (data['amount'] ?? 0).toDouble();
          }
        } else {
          amount = (data['amount'] ?? 0).toDouble();
        }

        return models.Order(
          id: doc.id,
          userId: userId,
          planName: planName,
          amount: amount,
          orderDate: (data['orderDate'] as Timestamp?)?.toDate() ?? DateTime.now(),
          status: models.OrderStatus.values.firstWhere(
            (e) => e.name == (data['status'] ?? 'pending'),
            orElse: () => models.OrderStatus.pending,
          ),
          billingCompleted: data['billingCompleted'] ?? false,
          phoneNumber: data['phoneNumber']?.toString(),
          simType: data['simType']?.toString() ?? '',
          currentStep: data['currentStep'] as int?,
        );
      }).toList();
    } catch (e) {
      print('DEBUG: Error fetching orders by status: $e');
      return [];
    }
  }

  /// Cancel an order by setting its status to "draft"
  Future<bool> cancelOrder(String userId, String orderId) async {
    print('DEBUG: FirebaseOrderManager.cancelOrder called with orderId: $orderId');

    try {
      await _firebaseManager.updateOrder(userId, orderId, {
        'status': 'draft',
        'updatedAt': FieldValue.serverTimestamp(),
      });

      print('DEBUG: FirebaseOrderManager - Successfully cancelled order $orderId');
      return true;
    } catch (e) {
      print('DEBUG: FirebaseOrderManager - Failed to cancel order: $e');
      return false;
    }
  }

  /// Delete an order
  Future<bool> deleteOrder(String userId, String orderId) async {
    print('DEBUG: FirebaseOrderManager.deleteOrder called with orderId: $orderId');

    try {
      await _firebaseManager.deleteOrder(userId, orderId);
      print('DEBUG: FirebaseOrderManager - Successfully deleted order $orderId');
      return true;
    } catch (e) {
      print('DEBUG: FirebaseOrderManager - Failed to delete order: $e');
      return false;
    }
  }
}

